cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewManager.java

```diff
@@ -215,6 +217,17 @@ public ApplicationId start(@Name("namespaceId") NamespaceId namespace, AppReques
       throw new IllegalStateException("Preview service is not running. Cannot start preview for " + programId);
     }
 
+
+    Map<String, Object> config = appRequest.getConfig();
+    if (config.get(Constants.Preview.PREVIEW_REQUEST_NUM_OF_RECORDS) instanceof Double) {
+      Double numOfRecords = (Double) config.get(Constants.Preview.PREVIEW_REQUEST_NUM_OF_RECORDS);
```

Comment:
`numOfRecordsPreview` config is very specific to app. Some other application can have its own configuration name for this config. So we cannot use it in the platform deterministically. We can

1. Add a new API in the `DataTracer` interface
```java
public interface DataTracer {
   ...
   int getMaximumInputRecords();
}
```
2. Implementation of `DataTracer` is `DefaultDataTracer` which is in app-fabric module. It can get CConfiguration instance through juice injection in `DefaultDataTracerFactory`.
3. `DataTracer` is available to the app for example [here](https://github.com/cdapio/cdap/blob/develop/cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/mapreduce/MapReduceBatchContext.java#L67)

We can get maximum input records here too using new DataTracer method.
```java
this.maxPreviewRecords = context.getDataTracer(stageSpec.getName()).getMaximumInputRecords();
```
4. `AbstractStageContext.getMaxPreviewRecords` method can be modified now to accept the integer limit (from DataTracer). AbstractStageContext.getMaxPreviewRecords can return minimum of this limit and config from the app or throw an exception when config from the app exceeds the limit from DataTracer.

Reply:
We discussed that we'll move this limit logic to the validateConfig stage of PipelineSpecGenerator. I am closing this PR